src/arch/src/lib.rs

```diff
@@ -63,43 +61,88 @@ impl From<GuestMemoryError> for Error {
 
 pub type Result<T> = std::result::Result<T, Error>;
 
-pub fn create_fdt<T: GuestMemory>(
-    cmdline: &str,
-    num_vcpus: u32,
-    guest_mem: &T,
-    fdt_load_offset: u64,
-    serial_addr: u64,
-    rtc_addr: u64,
-) -> Result<()> {
-    let mut fdt = FdtWriter::new()?;
-
-    // The whole thing is put into one giant node with s
-    // ome top level properties
-    let root_node = fdt.begin_node("")?;
-    fdt.property_u32("interrupt-parent", PHANDLE_GIC)?;
-    fdt.property_string("compatible", "linux,dummy-virt")?;
-    fdt.property_u32("#address-cells", 0x2)?;
-    fdt.property_u32("#size-cells", 0x2)?;
-
-    create_chosen_node(&mut fdt, cmdline)?;
-
-    let mem_size: u64 = guest_mem.iter().map(|region| region.len() as u64).sum();
-    create_memory_node(&mut fdt, mem_size)?;
-    create_cpu_nodes(&mut fdt, num_vcpus)?;
-    create_gic_node(&mut fdt, true, num_vcpus as u64)?;
-    create_serial_node(&mut fdt, serial_addr)?;
-    create_rtc_node(&mut fdt, rtc_addr)?;
-    create_timer_node(&mut fdt, num_vcpus)?;
-    create_psci_node(&mut fdt)?;
-    create_pmu_node(&mut fdt, num_vcpus)?;
-
-    fdt.end_node(root_node)?;
-    let fdt_final = fdt.finish()?;
-
-    let fdt_address = GuestAddress(AARCH64_PHYS_MEM_START + fdt_load_offset);
-    guest_mem.write_slice(fdt_final.as_slice(), fdt_address)?;
+#[derive(Default)]
+pub struct FdtBuilder<'a> {
+    cmdline: Option<&'a str>,
+    mem_size: Option<u64>,
+    num_vcpus: Option<u32>,
+    serial_console: Option<(u64, u64)>,
+    rtc: Option<(u64, u64)>,
+}
 
-    Ok(())
+impl<'a> FdtBuilder<'a> {
```

Comment:
We should be able to add some unit tests now that the code is structured this way. 

What we can check is at least the error cases (once we add them):
- creating an fdt without either of num_vcpus, mem, cmdline returns an error
- writing the fdt outside the reserved memory fails
We can also test that a "good FDT" with everything we need in it can be successfully built from the builder.

I am still thinking about a better way to test the internals of the FDT, but that might actually not be possible at this moment (i.e. checking somehow that the RTC is actually added to the FDT blob if the `with_rtc` function is called).